File: java8/src/main/java/com/shekhargulati/ninetynine_problems/_01_lists/P21.java
package com.shekhargulati.ninetynine_problems._01_lists;

import java.util.List;
import java.util.Map;

/**
 * (*) Insert an element at a given position into a list
 */
public class P21 {

    public static <T> List<T> insertAt(List<T> list, int pos, T t) throws IllegalArgumentException {
        if (pos < 1) {
            throw new IllegalArgumentException("pos can't be less than 1");
        }
        list.add((pos - 1), t);
        return list;
    }


    public static <T> List<T> insertAt_split(List<T> list, int pos, T t) throws IllegalArgumentException {
        if (pos < 1) {
            throw new IllegalArgumentException("pos can't be less than 1");
        }

        Map<Boolean, List<T>> split = P17.split(list, pos);
        List<T> first = split.get(true);
        List<T> second = split.get(false);
        first.add(pos - 1, t);

        first.addAll(second);
        return first;
    }


}

File: scala/src/main/scala/com/shekhargulati/ninetynine_problems/_01_lists/P21.scala
package com.shekhargulati.ninetynine_problems._01_lists

import scala.annotation.tailrec
import scala.collection.mutable

/**
  * (*) Insert an element at a given position into a list
  */
object P21 {

  def insertAt[T](list: List[T], pos: Int, t: T): List[T] = {
    if (pos < 1) {
      throw new IllegalArgumentException("position should be greater than 0")
    }
    val (first, second) = list.splitAt(pos - 1)
    (first :+ t) ++ second
  }

  def insertAt_inbuilt[T](list: List[T], pos: Int, t: T): List[T] = {
    if (pos < 1) {
      throw new IllegalArgumentException("position should be greater than 0")
    }
    val listBuffer: mutable.Buffer[T] = list.toBuffer
    listBuffer.insert(pos - 1, t)
    listBuffer.toList
  }

  def insertAt_recursion[T](list: List[T], pos: Int, t: T): List[T] = {
    @tailrec
    def insertAt(remaining: List[T], first: List[T], cur: Int): List[T] = remaining match {
      case xs if cur == pos => (first :+ t) ++ remaining
      case x :: xs => insertAt(xs, first :+ x, cur + 1)
      case _ => Nil
    }
    insertAt(list, List(), 1)
  }

}
